### DISCUSSION TITLE
Raw and reliable messages with Geckos

### DISCUSSION THREAD

**Discussion by panuchka** [geckosio/geckos.io#109] [Help / Questions] [answered]
Hi all!

Am I correct to assume that you can't send raw reliable messages from server to client? 👀 If this is true, would this be possible feature to create? 😎 

I'm currently struggling to combine Geckos with [Javelin](https://github.com/3mcd/javelin/tree/next) (ECS framework) that provides some protocol functions that only use ArrayBuffers for communication between client <-> server, but I would need to send some reliable initialization data from server to client.

**Comment by panuchka** [comment, marked as answer]
Could @yandeu share some light on this one? 🙂

**Reply by yandeu** [reply]
It is not a good idea to use geckos.io for reliable messages. The whole point of UDP is to send unreliable (hence fast) messages.

My "reliable" function is very primitive and should not be used as the default. But only for little things. If you often require reliable messages you should probably use http or WebSockets.

But back to your question. You can't send arrayBuffers as reliable using geckos.

I recommend: If these message are only for initializing stuff, send a normal http request.

**Reply by panuchka** [reply]
Yes thanks for the reply boss! Will use websockets or http requests. Just have to figure out how many requests I need to send reliably between the start and end points of the game session...

**Reply by iamcsharper** [reply]
> Yes thanks for the reply boss! Will use websockets or http requests. Just have to figure out how many requests I need to send reliably between the start and end points of the game session...

Ah man you can still implement your own reliability using ack mechanism.
Give each message server sends a unique ID, as soon as client receieves it, it sends back the ACK message for e.g. 0xFF 0X0 0X0 0X0 0X1 which means id=1 was confirmed to be receieved
the only thing that is up to you is to keep a buffer of undelivered messages
